ext/opcache/Optimizer/sccp.c

```diff
@@ -906,6 +911,12 @@ static inline int ct_eval_func_call(
 				return FAILURE;
 			}
 			/* pass */
+		} else if (zend_string_equals_literal(name, "dirname")) {
```

Comment:
1. On windows, this seems to be locale dependent. Not sure if that can be safely changed or if that applies to windows or alternate platforms capable of running php.exe?
2. **You should check that the value is `>= 1`, so that it doesn't throw**

```
php > dirname(__DIR__, 0);
Warning: dirname(): Invalid argument, levels must be >= 1 in php shell code on line 1
```

https://man7.org/linux/man-pages/man3/isalpha.3.html
```
       isalpha()
              checks for an alphabetic character; in the standard "C"
              locale, it is equivalent to (isupper(c) || islower(c)).  In
              some locales, there may be additional characters for which
              isalpha() is true—letters which are neither uppercase nor
              lowercase.
```

Is the implementation of dirname also locale dependent (`isalpha`), for a different reason? I'd hope that windows drives only allow a-z and A-Z, no matter what the system locale is, but I'm not familiar with all windows versions.

```c
ZEND_API size_t zend_dirname(char *path, size_t len)
{
	register char *end = path + len - 1;
	unsigned int len_adjust = 0;

#ifdef ZEND_WIN32
	/* Note that on Win32 CWD is per drive (heritage from CP/M).
	 * This means dirname("c:foo") maps to "c:." or "c:" - which means CWD on C: drive.
	 */
	if ((2 <= len) && isalpha((int)((unsigned char *)path)[0]) && (':' == path[1])) {
```